lib/smooth/include/smooth/core/util/DmaFixedBuffer.h

```diff
@@ -0,0 +1,89 @@
+/*
+Smooth - A C++ framework for embedded programming on top of Espressif's ESP-IDF
+Copyright 2019 Per Malmberg (https://gitbub.com/PerMalmberg)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// **************************** SPECIAL NOTE ***********************************************
+// This class was created with the goal of implementing the requirements for DMA as described
+// in ESP-IDF SPI Master API in section "Write and Read Phases".  This class should be useful
+// with periperals in the ESP32 chip that support hardware DMA; namely SPI and I2S.
+
+#pragma once
+
+#include <algorithm>
+#include <heap/include/esp_heap_caps.h>
+#include <smooth/core/logging/log.h>
+
+namespace smooth::core::util
+{
+    template<typename T, size_t Size>
+    class DmaFixedBuffer
+    {
+        public:
+            DmaFixedBuffer()
+            {
+                buff = static_cast<T*>(heap_caps_malloc(Size, MALLOC_CAP_DMA));
+
+                // Force the user to create DmaFixedBuffer in multiples of 4 bytes
+                static_assert(sizeof(T) * Size % 4 == 0,
+                              "DmaFixedBuffer must have a length of multiple of 32 bits, i.e. 4 bytes");
```

Comment:
Two things:
- This is evaluated at compile-time so mixing it with the run-time code just confuses things imho. I prefer to place them at the top of the class.
- While the multiples of 4 bytes applies to your specific use case, lets not make this class only usable for this particular usage, shall we? I suggest you do one of these just below `public:`:

```
constexpr size_t size() { return sizeof(T) * Size; }
```
That way it is possible to do a static assert on the buffer instance itself where it is used: `static_assert(buffer_instance::size() % 4 == 0, "blah blah")`

(I hope I got the syntax correct)